### ISSUE TITLE
Support WebdriverIO

### ISSUE THREAD

**Issue by jan-molak** [serenity-js/serenity-js#805]
In the light of angular/protractor#5502, Serenity/JS should provide an alternative to Protractor to enable interacting with Web applications.

Based on the [recent survey](https://twitter.com/SerenityJS/status/1379368915017592835) and discussion at #493, WebdriverIO seems like a clear winner.

This ticket lists all the features that the new `@serenity-js/webdriverio` module would need in order to become a drop-in replacement for `@serenity-js/protractor` module and allow for easy migration.

While WebdriverIO provides additional features such as [mocking](https://webdriver.io/docs/api/browser/mock/), [performance testing](https://webdriver.io/docs/devtools-service/#performance-testing), and so on, those are out of scope of this ticket (which is already massive).

## Help me help you

👋🏻  Hello, Fellow Developer! 

Introducing support for WebdriverIO is quite an undertaking, so if my work on Serenity/JS helps you get your work done, please consider becoming a [GitHub Sponsor](https://github.com/sponsors/serenity-js) of Serenity/JS. 

Thank you!
Jan

---

## Reporting

To provide reporting capabilities to WebdriverIO:
- [x] `@serenity-js/core`
  - [x] must drop support for Node 10.x - #842
- [x] `@serenity-js/webdriverio`
  - [x] must provide a WebdriverIO framework adapter to enable integration between WebdriverIO and Serenity/JS modules
  - [x] must emit [WebdriverIO events](https://github.com/webdriverio/webdriverio/tree/main/packages/wdio-reporter) to allow for any existing WebdriverIO reporters to be used with Serenity/JS framework adapter
  - [x] must tag scenarios to capture browser name, version, operating system
- [ ] `@serenity-js/cucumber`
  - [x] must integrate with `WebdriverIOFrameworkAdapter`
  - [x] could expose the number of spec files to be executed to WDIO
  - [ ] could have a configurable timeout, ([`wdio-cucumber-framework`](https://github.com/webdriverio/webdriverio/blob/a59dfead12c317e40dff1dc4001151796b25d030/packages/wdio-cucumber-framework/src/index.ts#L134))
  - [ ] Cucumber reporters could emit `TestSuiteStarts` and `TestSuiteFinished` (same as`@serenity-js/jasmine`) before and after feature starts/finishes, probably the same goes for `Rule`s
   - [x] should emit `RetryableSceneDetected` when a test scenario will be retried upon failure
- [x] `@serenity-js/mocha`
   - [x] must integrate with `WebdriverIOFrameworkAdapter`
   - [x] must expose the number of spec files to be executed to WDIO
   - [x] `SerenityReporterForMocha` should emit `TestSuiteStarts` and `TestSuiteFinished` (same as`@serenity-js/jasmine`) so that Serenity/JS framework adapter can emit `suite:start` and `suite:end` events required by WebdriverIO reporters
   - [x] should emit `RetryableSceneDetected` when a test scenario will be retried upon failure
- [x] `@serenity-js/jasmine`
  - [x] must integrate with `WebdriverIOFrameworkAdapter`
  - [x] must expose the number of spec files to be executed to WDIO
  - [x] should support `grep`, `invertGrep` and `specFilter` config options
- [ ] `@serenity-js/console-reporter` 
  - [x] must accept WebdriverIO's [`stdout` buffer](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-local-runner/src/index.ts#L18-L19) instead of using `process.stdout`
  - [x] must support `--watch` mode
  - [x] must support parallel runners
  - [ ] should print Serenity `Tag`s, like browser name and version, operating system, cucumber tags, etc.
  
## Screenplay 

Serenity/JS should provide a Screenplay Pattern abstraction over WebdriverIO `browser`, compatible where possible with `@serenity-js/protractor` to allow for easy migration.

### Abilities
- [x] BrowseTheWeb
  -  `BrowseTheWeb` should allow for a specific [multi-remote browser](https://webdriver.io/docs/multiremote/#) to be used, i.e. `actorA.whoCan(BrowseTheWeb.using(browser['browser-for-actorA'])`

### Interactions:
- [x] working with alerts
    - [x] Accept
    - [x] Dismiss
- [x] modifying browser config
    - [x] DeleteCookies
        - [x] ... maybe add interactions to add and modify cookies too while I'm at it
    - [x] ExecuteScript
- [x] working with forms
    - [x] Select
    - [x] Enter
      - [x] aggregate WebdriverIO special keys in something similar to `protractor.Key` enum
    - [x] Clear
- [x] working with frames and windows
    - [x] Switch
    - [x] Close
    - [x] ResizeBrowserWindow -> `Page#setViewportSize(w, h)` and `Page#viewportSize()`
- [x] navigating
    - [x] Navigate
      - [x] Navigate.to
      - [x] Navigate.back
      - [x] Navigate.forward
      - [x] Navigate.reloadPage
    - [x] Click
    - [x] DoubleClick
    - [x] RightClick
    - [x] Scroll
    - [x] Hover
    - [x] Wait
    - [x] TakeScreenshot

### Questions:
- [ ] by - a replacement of Protractor `by` that wraps [WebdriverIO selectors](https://webdriver.io/docs/selectors/)
  - [x] by.css
  - [x] by.tagName
  - [x] by.xpath
  - [x] by.linkText
  - [x] by.partialLinkText
  - [ ] by.role
  - [x] by.id
  - [ ] mobile selectors
  - [ ] react
- [x] Target
  - [x] single element
  - [x] multiple elements
  - [x] nested elements
  - [x] list operations (first, last, filter, etc.)
- [x] Text (Text.of, Text.ofAll)
- [x] Attribute
  - [x] **(protractor)** DSL changed to support using as a filter with lists; update Protractor implementation to match
- [x] ~Browser~ (removed; I'll need to re-think it)
- [x] Cookie
- [x] CSSClasses
- [x] LastScriptExecution
- [x] ModalDialog
- [x] Selected
- [x] Value
- [x] Website
- [x] Window -> replaced by `Page`

### Expectations

- [x] isActive
- [x] isClickable
- [x] isEnabled
- [x] isPresent
- [x] isSelected
- [x] isVisible

### Cast

- [x] Photographer 

## WebdriverIO CLI

To make it easier for folks to get started with Serenity/JS, WebdriverIO CLI should be extended to generate Serenity/JS-specific boilerplate code (requires PR to WebdriverIO). 

## Out of scope

- Support for `webdriverio/sync` (see https://github.com/webdriverio/webdriverio/discussions/6702)

**Comment by SJ1591** [comment]
Wow, pretty excited about it

**Comment by viper3400** [comment]
@jan-molak What are your plans with the `Target`s, so far? The `by` locator is directly imported from the `protractor` module itself. Do you plan to adapt this pattern in a generic way in order to minimize migration effort?

**Comment by jan-molak** [comment]
Hey @viper3400!

> @jan-molak What are your plans with the Targets, so far? The by locator is directly imported from the protractor module itself. Do you plan to adapt this pattern in a generic way in order to minimize migration effort?

That's a good question and I'm open to suggestions. My thoughts so far are that WebdriverIO has recently introduced [better support for shadow dom](https://webdriver.io/docs/selectors/#deep-selectors), so it seems like most if not all selectors could be expressed as `string`.

So perhaps what we need is something like:
```
Target.the('widget').locatedBy('string selector')
```

For WebdriverIO this would delegate to [`$('string selector')`](https://webdriver.io/docs/selectors/) and for Protractor it could be an alias for `Target.the('widget').located(by.css('string selector')`.

An alternative design would be to introduce a Serenity/JS-specific `by` locators for WebdriverIO, which would do the delegation.

While the first approach feels simpler, the second one offers more flexibility since WDIO offers selectors like `react$`, which I don't think can be expressed as `string` only (is that right, @christian-bromann?).

Thinking about it, we could probably do both too.

**Comment by christian-bromann** [comment]
I would advocate to write a [codemod](https://github.com/webdriverio/codemod) for it. We use it to help users already migrate from Protractor to WebdriverIO and it already has transformation rules implemented to transform all these locators into WebdriverIO ones.

**Comment by viper3400** [comment]
This is my point, I don't want to migrate from one locator style to another. I assume, most tools implement at least something like the CSS selectors. This seems a generic and tool agnostic approach to me. 

Sure, if you have to rely on special features, you will have to stick to the special module (which is in fact the case with the `by` locator today.)

**Comment by toniisidoro-sofist** [comment]
Hello guys I want to help in any way with this integration with webdriver.io but I'm in doubt about how to help and where to start. Maybe I can start another integration with Cypress or Playwright in parallel while you guys are focused on this one. What do you think?

**Comment by jan-molak** [comment]
@toniisidoro-sofist awesome, thanks for offering to help! It would be great to get started on investigating integration with Cypress. I think `@serenity-js/mocha` reporter _should be_ able to support it with some minor tweaks, but I didn't have a chance to have a proper play with it yet. If that's something you'd be interested in picking up we could move that conversation to a separate ticket perhaps?

**Comment by viper3400** [comment]
@jan-molak, after nearly a year of work your've now reached 3.0.0-rc.13. 

What are the plans for an official release of 3.0.0? 
 - Which major features are still missing? 
 - Are there blocking bugs?

I wonder, if we shall focus on getting the release stable and published?

I'm pretty sure, there is still some work to do to out of feature implementation:
- migrate templates, 
- edit docs at serenity-js.org

I wonder, if it would it be possible to keep the 2.x docs online for a while, once there is a 3.x release?

**Comment by jan-molak** [comment]
> What are the plans for an official release of 3.0.0?

I'd like to stabilise, document, and provide examples for any "popular" public APIs. Turns out the last 10% of the project takes 90% of the time ;-)

> Which major features are still missing?

There's nothing specific missing, just a couple of things I need to figure out:
- CSS selectors
  - I'm strongly considering limiting `By.*` selectors to `By.css` and `By.xpath`, and removing selectors that feel outdated and redundant, e.g. `By.tagName`, `By.cssContainingText`,  and `By.id`. `By.id` and `By.tagName` could be easily replaced by `By.css`, and `By.linkContainingText` is easy to replace by filter queries (`PageElements...where(Text, equals(..))`).
  - It is a bit of a drastic move, so I'm weighing the pros and the cons. It's a very popular public API, so changing my mind about its design after the API is published and incorporated into people's test suites bears a risk of breaking changes.
  - given the popularity of Web components, I'm also considering if `By.css` should also support shadow DOM by default. It seems like a plausible design, but there are [certain issues](https://github.com/Georgegriff/query-selector-shadow-dom#known-issues) I need to understand the implications of better. This is something I think we should consider post-3.0
- Notepad APIs
  - I'm finishing chapter 13 of [BDD in Action, second edition](https://www.manning.com/books/bdd-in-action-second-edition), which is dedicated to Serenity/JS. There are several examples there that rely on the ability to `TakeNotes`, so I need to review and clean up those APIs a bit to make sure I don't change them too drastically right after the book is published.

> Are there blocking bugs?

It feels like we've managed to address all of them at the moment. It would be nice to address #1147, but if we can't then we can do it after 3.0.

> I wonder, if we shall focus on getting the release stable and published?

Absolutely! 

> I'm pretty sure, there is still some work to do to out of feature implementation:

Yes, it's not too much, but needs to be done :)

> migrate templates,

This is something I think we could actually start straight away (and, in fact, @gigamac has already started in https://github.com/serenity-js/serenity-js-cucumber-template/pull/209).

What I think we'd need to do is:
- create a branch called `main` in each template (based on the current `master`)
- migrate the code to use `3.0.0-rc`
- when 3.0 is published, swap the default branch in each template to `main` and rename `master` to `2.x`

🎫 It feels like migrating each template is a nice, isolated piece of work that someone could pick up.

> edit docs at serenity-js.org

The API docs will generate themselves, but we should revisit all the [articles in the "Handbook"](https://github.com/serenity-js/serenity-js/tree/main/documentation/website/src/handbook)
There's nothing blocking it, I just didn't have any capacity to look into it yet...

🎫 If anyone has the capacity to look into it, that would make for a nice isolated piece of work.

I don't expect too much of a problem there either, we'll need to go through [each article](https://github.com/serenity-js/serenity-js/tree/main/documentation/website/src/handbook) and update the code samples since the rest of the content should still hold (I think).

> I wonder, if it would it be possible to keep the 2.x docs online for a while, once there is a 3.x release?

⚠️ It would be nice, but we'd need to find a sensible way to do it. ESDoc has been quite a challenge to get it to work with a mono-repo, so I'd rather not touch it again unless I really had to (to be fair, it works great for simple, single-module projects, for which it's been designed).

I'm actually considering migrating the website to another static website generator like Docusaurus or something similar, but that's a project on its own. I'd rather do it after 3.0

**Comment by viper3400** [comment]
OK, to keep track of it I suppose to create some other issues:
- one for the docs at serenity-js.org and the details
- one for each template that have to be migrated (not sure, if this should happen in their dedicated repository issue or here. Having it here would be more trackable)
- one for each thing you’d like to have in 3.0 (if not still exists)

Further, we can think about creating a new milestone 3.1 and move all things not necessary for 3.0 to 3.1.

I’d rather move #1147 to a later release, as it’s currently blocked by the not yet merged PR. 
I’d even say, that 3.0 can go without having all the templates migrated, maybe we can prioritise some.
I could have a look into the docs, if you’d want me to.

**Comment by jan-molak** [comment]
> OK, to keep track of it I suppose to create some other issues:

Sounds great, thanks!

> Further, we can think about creating a new milestone 3.1 and move all things not necessary for 3.0 to 3.1

Agreed

> I’d rather move https://github.com/serenity-js/serenity-js/issues/1147 to a later release, as it’s currently blocked by the not yet merged PR.

I think the fix might take some time to implement on the Serenity BDD CLI side since it's not trivial.

> I’d even say, that 3.0 can go without having all the templates migrated, maybe we can prioritise some.

Yes, I think WDIO Cucumber an WDIO Mocha are probably the most important ones

> I could have a look into the docs, if you’d want me to.

I'd be grateful, thank you!

**Comment by jan-molak** [comment]
🚀 All the public APIs are now stable in 3.0.0-rc

To get started, use the project templates, `main` branch:
- https://github.com/serenity-js/serenity-js-mocha-webdriverio-template/tree/main
- https://github.com/serenity-js/serenity-js-cucumber-webdriverio-template/tree/main

New tutorials coming soon; join [Serenity/JS Community Chat](https://gitter.im/serenity-js/Lobby) and follow [@SerenityJS](https://twitter.com/SerenityJS) on Twitter to get notified when they arrive!
